src/Microsoft.AspNetCore.Server.HttpSys/AuthenticationHandler.cs

```diff
@@ -16,111 +16,94 @@ internal class AuthenticationHandler : IAuthenticationHandler
-                                context.Accept();
-                            }
-                            else
-                            {
-                                _requestContext.Response.StatusCode = 401;
-                                _customChallenges |= scheme;
-                                context.Accept();
+                                break;
                             }
-                            break;
-                        default:
-                            throw new NotSupportedException(context.Behavior.ToString());
+                        }
                     }
-                }
+                    else
+                    {
+                        _requestContext.Response.StatusCode = 401;
+                        foreach (var scheme in ListEnabledAuthSchemes())
+                        {
+                            _customChallenges |= scheme;
+                        }
+                    }
+                    break;
+                default:
+                    throw new NotSupportedException(context.Behavior.ToString());
             }
+
             // A challenge was issued, it overrides any pre-set auth types.
             _requestContext.Response.AuthenticationChallenges = _customChallenges;
             return TaskCache.CompletedTask;
         }
 
-        public void GetDescriptions(DescribeSchemesContext context)
+        public Task InitializeAsync(AuthenticationScheme scheme, HttpContext context)
         {
-            // TODO: Caching, this data doesn't change per request.
-            foreach (var scheme in ListEnabledAuthSchemes())
+            _scheme = scheme;
+            _requestContext = context.Features.Get<RequestContext>();
+
+            if (_requestContext == null)
             {
-                context.Accept(GetDescription(scheme.ToString()));
+                throw new InvalidOperationException("No RequestContext found.");
             }
+
+            _authSchemes = _requestContext.Response.AuthenticationChallenges;
+            _customChallenges = AuthenticationSchemes.None;
+            return TaskCache.CompletedTask;
         }
 
         public Task SignInAsync(SignInContext context)
         {
-            // Not supported. AuthenticationManager will throw if !Accepted.
-            return TaskCache.CompletedTask;
+            throw new NotSupportedException();
         }
 
         public Task SignOutAsync(SignOutContext context)
         {
-            // Not supported. AuthenticationManager will throw if !Accepted.
             return TaskCache.CompletedTask;
         }
 
-        private IDictionary<string, object> GetDescription(string authenticationScheme)
-        {
-            return new Dictionary<string, object>()
-            {
-                { "AuthenticationScheme", authenticationScheme },
-            };
-        }
-
         private IEnumerable<AuthenticationSchemes> ListEnabledAuthSchemes()
```

Comment:
I don't think you need ListEnabledAuthSchemes anywhere now that you only have one handler for all schemes.